sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -176,10 +176,13 @@ abstract class UnaryNode extends LogicalPlan {
         allConstraints += EqualNullSafe(a.toAttribute, l)
       case a @ Alias(e, _) =>
         // For every alias in `projectList`, replace the reference in constraints by its attribute.
-        allConstraints ++= allConstraints.map(_ transform {
-          case expr: Expression if expr.semanticEquals(e) =>
-            a.toAttribute
-        })
+        allConstraints ++= allConstraints.map {
+          case e @ EqualNullSafe(l, _: AttributeReference) if l.references.size > 1 => e
```

Comment:
I feel its a bit difficult to understand this pattern-matching at a glance, could you leave comments about what this means? Probably, you wanna skip [the pattern below](https://github.com/apache/spark/pull/26257/files#diff-27c76f96a7b2733ecfd6f46a1716e153R186) for performance?